src/conversions/array.rs

```diff
@@ -149,17 +174,69 @@ mod min_const_generics {
 #[cfg(not(min_const_generics))]
 mod array_impls {
     use super::invalid_sequence_length;
-    use crate::{FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python, ToPyObject};
+    use crate::conversion::IntoPyPointer;
+    use crate::{
+        ffi, FromPyObject, IntoPy, Py, PyAny, PyObject, PyResult, PyTryFrom, Python, ToPyObject,
+    };
+    use std::mem::{transmute_copy, ManuallyDrop};
 
     macro_rules! array_impls {
         ($($N:expr),+) => {
             $(
                 impl<T> IntoPy<PyObject> for [T; $N]
                 where
-                    T: ToPyObject
+                    T: IntoPy<PyObject>
                 {
                     fn into_py(self, py: Python<'_>) -> PyObject {
-                        self.as_ref().to_object(py)
+
+                        struct ArrayGuard<T> {
+                            elements: [ManuallyDrop<T>; $N],
+                            start: usize,
+                        }
+
+                        impl<T> Drop for ArrayGuard<T> {
+                            fn drop(&mut self) {
+                                unsafe {
+                                    let needs_drop = self.elements.get_mut(self.start..).unwrap();
+                                    for item in needs_drop{
+                                        ManuallyDrop::drop(item);
+                                    }
+                                }
+                            }
+                        }
+
+                        unsafe {
+                            let ptr = ffi::PyList_New($N as ffi::Py_ssize_t);
+
+                            // We create the  `Py` pointer here for two reasons:
+                            // - panics if the ptr is null
+                            // - its Drop cleans up the list if user code panics.
+                            let list: Py<PyAny> = Py::from_owned_ptr(py, ptr);
+
+                            let slf = ManuallyDrop::new(self);
+
+                            let mut guard = ArrayGuard{
+                                // the transmute size check is _very_ dumb around generics
+                                elements: transmute_copy(&slf),
```

Comment:
I'm curious why the `ManuallyDrop::new` followed by `transmute_copy`, instead of just `elements: transmute(self)`?

Reply:
It doesn't work - you have to use `transmute_copy` because it doesn't have `transmute`'s size check. For example...

```rust
fn into<const N: usize>(input: [u8; N]) -> [i8; N] {
    unsafe { std::mem::transmute(input) }
}
```
...does not work.
```
 --> src\main.rs:8:14
  |
8 |     unsafe { std::mem::transmute(input) }
  |              ^^^^^^^^^^^^^^^^^^^
  |
  = note: source type: `[u8; N]` (this type does not have a fixed size)
  = note: target type: `[i8; N]` (this type does not have a fixed size)
```

Reply:
Hmmm, is that a compiler bug? An array has a fixed size 🤔

Either way, makes sense, let's stick with `transmute_copy`!